In [1]:
import pandas as pd
pd.set_option('display.max_columns',None)
pd.set_option('display.max_rows',None)

In [2]:
application = pd.read_csv(r"C:\Users\meet.gajera\OneDrive\Desktop\hcdr__\Data\featured\application_featured.csv")
bureau = pd.read_csv(r"C:\Users\meet.gajera\OneDrive\Desktop\hcdr__\Data\featured\bureau_featured.csv")

In [3]:
application.head()

,SK_ID_CURR,TARGET,CODE_GENDER,CNT_CHILDREN,CNT_FAM_MEMBERS,NAME_CONTRACT_TYPE,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,AMT_GOODS_PRICE,NAME_INCOME_TYPE,OCCUPATION_TYPE,OCCUPATION_TYPE_MISSING,NAME_EDUCATION_TYPE,NAME_FAMILY_STATUS,FLAG_OWN_CAR,FLAG_OWN_REALTY,NAME_HOUSING_TYPE,FLAG_EMAIL,FLAG_WORK_PHONE,AGE,YEARS_EMPLOYED,YEARS_CURRENT_PHONE,ORG_GROUP,CREDIT_INCOME_RATIO,ANNUITY_INCOME_RATIO,CREDIT_ANNUITY_RATIO,GOODS_CREDIT_RATIO,INCOME_PER_FAM_MEMBER,EMPLOYMENT_AGE_RATIO,IS_SINGLE_PARENT,HAS_DEPENDENTS
0,100002,1,M,0,1.0,Cash loans,202500.0,406597.5,24700.5,351000.0,Working,Laborers,0,Secondary / secondary special,Single / not married,N,Y,House / apartment,0,0,25,1.75,3.11,Business & Corporate,2.0079,0.1220,16.4611,0.8633,202500.0,0.0700,0,0
1,100003,0,F,0,2.0,Cash loans,270000.0,1293502.5,35698.5,1129500.0,State servant,Core staff,0,Higher education,Married,N,N,House / apartment,0,0,45,3.25,2.27,Education,4.7908,0.1322,36.2341,0.8732,135000.0,0.0722,0,0
2,100004,0,M,0,1.0,Revolving loans,67500.0,135000.0,6750.0,135000.0,Working,Laborers,0,Secondary / secondary special,Single / not married,Y,Y,House / apartment,0,1,52,0.62,2.23,Government,2.0000,0.1000,20.0000,1.0000,67500.0,0.0119,0,0
3,100006,0,F,0,2.0,Cash loans,135000.0,312682.5,29686.5,297000.0,Working,Laborers,0,Secondary / secondary special,Civil marriage,N,Y,House / apartment,0,0,52,8.33,1.69,Business & Corporate,2.3162,0.2199,10.5328,0.9498,67500.0,0.1602,0,0
4,100007,0,M,0,1.0,Cash loans,121500.0,513000.0,21865.5,513000.0,Working,Core staff,0,Secondary / secondary special,Single / not married,N,Y,House / apartment,0,0,54,8.32,3.03,Other,4.2222,0.1800,23.4616,1.0000,121500.0,0.1541,0,0


In [4]:
bureau.head()

,SK_ID_CURR,AMT_CREDIT_SUM_sum,AMT_CREDIT_SUM_mean,AMT_CREDIT_SUM_max,AMT_CREDIT_SUM_DEBT_sum,AMT_CREDIT_SUM_DEBT_mean,AMT_CREDIT_SUM_DEBT_MISSING_sum,AMT_CREDIT_SUM_DEBT_MISSING_mean,AMT_CREDIT_SUM_LIMIT_sum,AMT_CREDIT_SUM_LIMIT_MISSING_sum,AMT_CREDIT_SUM_LIMIT_MISSING_mean,DEBT_RATIO_mean,DEBT_RATIO_max,CREDIT_UTILIZATION_mean,HAS_DEBT_sum,IS_ACTIVE_sum,IS_CLOSED_sum,IS_SOLD_sum,IS_BAD_DEBT_sum,YEARS_CREDIT_min,YEARS_CREDIT_max,YEARS_CREDIT_mean
0,100001,1453365.000,207623.571429,378000.0,596686.5,85240.928571,0,0.000,0.000,1,0.142857,0.282514,0.9874,207623.571429,3,3,4,0,0,0.134247,4.306849,2.013699
1,100002,865055.565,108131.945625,450000.0,245781.0,30722.625000,3,0.375,31988.565,4,0.500000,0.068275,0.5462,104133.500000,1,2,6,0,0,0.282192,3.936986,2.394521
2,100003,1017400.500,254350.125000,810000.0,0.0,0.000000,0,0.000,810000.000,0,0.000000,0.000000,0.0000,51850.375000,0,1,3,0,0,1.660274,7.084932,3.837671
3,100004,189037.800,94518.900000,94537.8,0.0,0.000000,0,0.000,0.000,0,0.000000,0.000000,0.0000,94518.900000,0,0,2,0,0,1.117808,3.632877,2.375342
4,100005,657126.000,219042.000000,568800.0,568408.5,189469.500000,0,0.000,0.000,0,0.000000,0.601233,0.9548,219042.000000,2,2,1,0,0,0.169863,1.021918,0.522374


In [5]:
application['SK_ID_CURR'].isin(bureau['SK_ID_CURR']).mean() * 100

85.6847397328876

confirms that about 85.68% of applicants have bureau information and about 14.32% do not.

In [6]:
merge_df1 = application.merge(
    bureau,
    on='SK_ID_CURR',
    how='left'
)

In [7]:
merge_df1.isnull().sum().sort_values(ascending=False)

YEARS_CREDIT_mean                    44021
DEBT_RATIO_mean                      44021
AMT_CREDIT_SUM_sum                   44021
AMT_CREDIT_SUM_mean                  44021
AMT_CREDIT_SUM_max                   44021
AMT_CREDIT_SUM_DEBT_sum              44021
AMT_CREDIT_SUM_DEBT_mean             44021
AMT_CREDIT_SUM_DEBT_MISSING_sum      44021
AMT_CREDIT_SUM_DEBT_MISSING_mean     44021
AMT_CREDIT_SUM_LIMIT_sum             44021
AMT_CREDIT_SUM_LIMIT_MISSING_mean    44021
AMT_CREDIT_SUM_LIMIT_MISSING_sum     44021
DEBT_RATIO_max                       44021
CREDIT_UTILIZATION_mean              44021
HAS_DEBT_sum                         44021
IS_ACTIVE_sum                        44021
IS_CLOSED_sum                        44021
IS_SOLD_sum                          44021
IS_BAD_DEBT_sum                      44021
YEARS_CREDIT_min                     44021
YEARS_CREDIT_max                     44021
GOODS_CREDIT_RATIO                       0
HAS_DEPENDENTS                           0
IS_SINGLE_P

## Post-Merge Missing Value Handling & Feature Engineering

In [8]:
# bureau history indicator

merge_df1['HAS_BUREAU_HISTORY'] = (
    merge_df1['AMT_CREDIT_SUM_sum'].notna().astype(int)
)

merge_df1.insert(
    merge_df1.columns.get_loc('AMT_CREDIT_SUM_sum'),
    'HAS_BUREAU_HISTORY',
    merge_df1.pop('HAS_BUREAU_HISTORY')
)

In [9]:
# Count features

count_cols = [
    'HAS_DEBT_sum',
    'IS_ACTIVE_sum',
    'IS_CLOSED_sum',
    'IS_SOLD_sum',
    'IS_BAD_DEBT_sum',
    'AMT_CREDIT_SUM_DEBT_MISSING_sum',
    'AMT_CREDIT_SUM_LIMIT_MISSING_sum'
]

merge_df1[count_cols] = merge_df1[count_cols].fillna(0)

In [10]:
# Amount features

amount_cols = [
    'AMT_CREDIT_SUM_sum',
    'AMT_CREDIT_SUM_mean',
    'AMT_CREDIT_SUM_max',
    'AMT_CREDIT_SUM_DEBT_sum',
    'AMT_CREDIT_SUM_DEBT_mean',
    'AMT_CREDIT_SUM_LIMIT_sum'
]

merge_df1[amount_cols] = merge_df1[amount_cols].fillna(0)

In [11]:
#  Ratio features

ratio_cols = [
    'DEBT_RATIO_mean',
    'DEBT_RATIO_max',
    'CREDIT_UTILIZATION_mean'
]

merge_df1[ratio_cols] = merge_df1[ratio_cols].fillna(0)

In [12]:
# Time features

time_cols = [
    'YEARS_CREDIT_min',
    'YEARS_CREDIT_mean',
    'YEARS_CREDIT_max'
]

for col in time_cols:
    merge_df1[col] = merge_df1[col].fillna(merge_df1[col].median())

`YEARS_CREDIT` features represent the age of a customer's credit history. Filling missing values with `0` would incorrectly imply that the customer obtained credit today (0 years ago). Therefore, missing values were imputed using the median to preserve a realistic representation of credit age.

In [13]:
# missing indicators

merge_df1[
    ['AMT_CREDIT_SUM_DEBT_MISSING_mean',
        'AMT_CREDIT_SUM_LIMIT_MISSING_mean']
] = merge_df1[
    ['AMT_CREDIT_SUM_DEBT_MISSING_mean',
        'AMT_CREDIT_SUM_LIMIT_MISSING_mean']
].fillna(0)

In [17]:
print(merge_df1.shape)
print(merge_df1.isnull().sum().sort_values(ascending=False).head(20))
print(merge_df1.duplicated().sum())

(307511, 54)
SK_ID_CURR                           0
AMT_CREDIT_SUM_LIMIT_sum             0
EMPLOYMENT_AGE_RATIO                 0
IS_SINGLE_PARENT                     0
HAS_DEPENDENTS                       0
HAS_BUREAU_HISTORY                   0
AMT_CREDIT_SUM_sum                   0
AMT_CREDIT_SUM_mean                  0
AMT_CREDIT_SUM_max                   0
AMT_CREDIT_SUM_DEBT_sum              0
AMT_CREDIT_SUM_DEBT_mean             0
AMT_CREDIT_SUM_DEBT_MISSING_sum      0
AMT_CREDIT_SUM_DEBT_MISSING_mean     0
AMT_CREDIT_SUM_LIMIT_MISSING_sum     0
TARGET                               0
AMT_CREDIT_SUM_LIMIT_MISSING_mean    0
DEBT_RATIO_mean                      0
DEBT_RATIO_max                       0
CREDIT_UTILIZATION_mean              0
HAS_DEBT_sum                         0
dtype: int64
0


In [15]:
merge_df1.head()

,SK_ID_CURR,TARGET,CODE_GENDER,CNT_CHILDREN,CNT_FAM_MEMBERS,NAME_CONTRACT_TYPE,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,AMT_GOODS_PRICE,NAME_INCOME_TYPE,OCCUPATION_TYPE,OCCUPATION_TYPE_MISSING,NAME_EDUCATION_TYPE,NAME_FAMILY_STATUS,FLAG_OWN_CAR,FLAG_OWN_REALTY,NAME_HOUSING_TYPE,FLAG_EMAIL,FLAG_WORK_PHONE,AGE,YEARS_EMPLOYED,YEARS_CURRENT_PHONE,ORG_GROUP,CREDIT_INCOME_RATIO,ANNUITY_INCOME_RATIO,CREDIT_ANNUITY_RATIO,GOODS_CREDIT_RATIO,INCOME_PER_FAM_MEMBER,EMPLOYMENT_AGE_RATIO,IS_SINGLE_PARENT,HAS_DEPENDENTS,HAS_BUREAU_HISTORY,AMT_CREDIT_SUM_sum,AMT_CREDIT_SUM_mean,AMT_CREDIT_SUM_max,AMT_CREDIT_SUM_DEBT_sum,AMT_CREDIT_SUM_DEBT_mean,AMT_CREDIT_SUM_DEBT_MISSING_sum,AMT_CREDIT_SUM_DEBT_MISSING_mean,AMT_CREDIT_SUM_LIMIT_sum,AMT_CREDIT_SUM_LIMIT_MISSING_sum,AMT_CREDIT_SUM_LIMIT_MISSING_mean,DEBT_RATIO_mean,DEBT_RATIO_max,CREDIT_UTILIZATION_mean,HAS_DEBT_sum,IS_ACTIVE_sum,IS_CLOSED_sum,IS_SOLD_sum,IS_BAD_DEBT_sum,YEARS_CREDIT_min,YEARS_CREDIT_max,YEARS_CREDIT_mean
0,100002,1,M,0,1.0,Cash loans,202500.0,406597.5,24700.5,351000.0,Working,Laborers,0,Secondary / secondary special,Single / not married,N,Y,House / apartment,0,0,25,1.75,3.11,Business & Corporate,2.0079,0.1220,16.4611,0.8633,202500.0,0.0700,0,0,1,865055.565,108131.945625,450000.0,245781.0,30722.625,3.0,0.375,31988.565,4.0,0.5,0.068275,0.5462,104133.500,1.0,2.0,6.0,0.0,0.0,0.282192,3.936986,2.394521
1,100003,0,F,0,2.0,Cash loans,270000.0,1293502.5,35698.5,1129500.0,State servant,Core staff,0,Higher education,Married,N,N,House / apartment,0,0,45,3.25,2.27,Education,4.7908,0.1322,36.2341,0.8732,135000.0,0.0722,0,0,1,1017400.500,254350.125000,810000.0,0.0,0.000,0.0,0.000,810000.000,0.0,0.0,0.000000,0.0000,51850.375,0.0,1.0,3.0,0.0,0.0,1.660274,7.084932,3.837671
2,100004,0,M,0,1.0,Revolving loans,67500.0,135000.0,6750.0,135000.0,Working,Laborers,0,Secondary / secondary special,Single / not married,Y,Y,House / apartment,0,1,52,0.62,2.23,Government,2.0000,0.1000,20.0000,1.0000,67500.0,0.0119,0,0,1,189037.800,94518.900000,94537.8,0.0,0.000,0.0,0.000,0.000,0.0,0.0,0.000000,0.0000,94518.900,0.0,0.0,2.0,0.0,0.0,1.117808,3.632877,2.375342
3,100006,0,F,0,2.0,Cash loans,135000.0,312682.5,29686.5,297000.0,Working,Laborers,0,Secondary / secondary special,Civil marriage,N,Y,House / apartment,0,0,52,8.33,1.69,Business & Corporate,2.3162,0.2199,10.5328,0.9498,67500.0,0.1602,0,0,0,0.000,0.000000,0.0,0.0,0.000,0.0,0.000,0.000,0.0,0.0,0.000000,0.0000,0.000,0.0,0.0,0.0,0.0,0.0,0.821918,5.005479,2.876712
4,100007,0,M,0,1.0,Cash loans,121500.0,513000.0,21865.5,513000.0,Working,Core staff,0,Secondary / secondary special,Single / not married,N,Y,House / apartment,0,0,54,8.32,3.03,Other,4.2222,0.1800,23.4616,1.0000,121500.0,0.1541,0,0,1,146250.000,146250.000000,146250.0,0.0,0.000,0.0,0.000,0.000,0.0,0.0,0.000000,0.0000,146250.000,0.0,0.0,1.0,0.0,0.0,3.147945,3.147945,3.147945


In [16]:
merge_df1.to_csv(r"C:\Users\meet.gajera\OneDrive\Desktop\hcdr__\Data\merged\merged1.csv")